### PR TITLE
[Transform] Handle multi-fields properly when creating destination index.

### PR DESCRIPTION
**Pull request by przemekwitek**
When source index contained a text & keyword field `foo`, creation of the destination index finished with error:
```
org.elasticsearch.ElasticsearchException: Elasticsearch exception [type=illegal_argument_exception, reason=can't merge a non object mapping [foo] with an object mapping]
```

This PR solves the problem by producing a proper mapping for a multi-field before creating the destination index.

Relates https://github.com/elastic/elasticsearch/issues/65869

Marking `>non-issue` as this fix is for an unreleased feature.

### CODE REVIEW

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/persistence/TransformIndex.java [hunk 1]
```diff
@@ -134,10 +134,28 @@ private static Settings createSettings() {
...
+    static Map<String, Object> createMappingsFromStringMap(Map<String, String> mappings) {
         Map<String, Object> fieldMappings = new HashMap<>();
-        mappings.forEach((k, v) -> fieldMappings.put(k, Map.of("type", v)));
-
+        for (Map.Entry<String, String> entry : mappings.entrySet()) {
+            String[] parts = entry.getKey().split("\\.");
+            Map<String, Object> current = fieldMappings;
+            current = diveInto(current, parts[0]);
+            for (int j = 1; j < parts.length; ++j) {
+                current = diveInto(current, "fields");
```
Comment (przemekwitek) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/persistence/TransformIndex.java:144]: Tests show this doesn't work when there are object fields mixed with multi-fields.
I need to rethink this solution.

**Comment by przemekwitek** [comment]
run elasticsearch-ci/default-distro

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/persistence/TransformIndex.java (hunk 1 above)
Comment (przemekwitek) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/persistence/TransformIndex.java]: Ok, I've fixed it.
Still, I have the impression that there should be an easier way of reconciling individual mappings into the new index creation request. But that's what I have for now.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/persistence/TransformIndex.java [hunk 2]
```diff
@@ -134,10 +148,28 @@ private static Settings createSettings() {
...
-    private static Map<String, Object> createMappingsFromStringMap(Map<String, String> mappings) {
+    static Map<String, Object> createMappingsFromStringMap(Map<String, String> mappings) {
+        List<Map.Entry<String, String>> sortedMappingsEntries = new ArrayList<>(mappings.entrySet());
+        // We sort the entry list to make sure that for each (parent, parent.child) pair, parent entry will be processed before child entry.
+        sortedMappingsEntries.sort(comparingByKey());
         Map<String, Object> fieldMappings = new HashMap<>();
-        mappings.forEach((k, v) -> fieldMappings.put(k, Map.of("type", v)));
-
+        for (Map.Entry<String, String> entry : sortedMappingsEntries) {
+            String[] parts = entry.getKey().split("\\.");
```
Comment (hendrikmuhs) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/persistence/TransformIndex.java:157]: nit: I am not a big fan of using regex if there is no reason. Arguably its java's fault. A potential alternative: the tokenize methods in `org.elasticsearch.common.Strings`

**Review by hendrikmuhs — APPROVED** [review]
LGTM

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/persistence/TransformIndex.java (hunk 2 above)
Comment (przemekwitek) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/persistence/TransformIndex.java:157]: > Arguably its java's fault.

Unfortunately I couldn't find a non-regex version of `split()` method.

> A potential alternative: the tokenize methods in org.elasticsearch.common.Strings

Thanks! I've used `Strings.tokenizeToStringArray` method.

**Pull request merged**